### ISSUE TITLE
Make suggestions when argument unrecognized

### ISSUE THREAD

**Issue by natemcmaster** [natemcmaster/CommandLineUtils#10]
Imitate some of git's awesomeness

![image](https://cloud.githubusercontent.com/assets/2696087/22569523/b506b886-e94c-11e6-8d2d-039b850410aa.png)

**Comment by natemcmaster** [comment]
Labeling as help-wanted. Let me know if you want to tackle it. Let's create a design first.

**Comment by MadbHatter** [comment]
I would like to tackle this, but I am not sure how.
Maybe add a Property like app.CommandSuggestion?
And edit the behaviour of HandleUnexpectedArg?

e.g. instead of:
```
if (_currentCommand.ThrowOnUnexpectedArgument)
{
    _currentCommand.ShowHint();
    throw new CommandParsingException(_currentCommand, $"Unrecognized {argTypeName} '{_enumerator.Current.Raw}'");
}
```
```
if (_currentCommand.ThrowOnUnexpectedArgument)
{
    if(_currentCommand.CommandSuggestion){
        //Calculate Levenshtein Distance and if close enough display hint?        
    }
    _currentCommand.ShowHint();
    throw new CommandParsingException(_currentCommand, $"Unrecognized {argTypeName} '{_enumerator.Current.Raw}'");
}
```

I guess I will think this over and most importantly get some shuteye/sleep.

**Comment by MadbHatter** [comment]
> What do you think of this?

Seems good to me.
Should we show only the best match or the best matches (If there are for some reason really similar commands/options)?

> As far as which algorithm and which threshold to select, i'm not sure. Requires more investigation. There are half a dozen good algorithms for string comparison, including Levenshtein distance. I'd have to play with them first to see what kind of suggestions they would make

I will see If I can't whip a short Demo for different String Comparsion Algorithms and Tresholds.

**Comment by MadbHatter** [comment]
This is not representative.
Seeing as
A: The mistypes are not natural at all.
B: I used only the standard settings (aka standard constructor) and I did not try to tweak the algorithms (or read up how they all work).

This is only meant as a Showcase and uses the Algorithms from SimMetrics:

Mangled Strings:

 Input| Output
------------ | -------------
config|conbig
init|iGit
clone|cWone\b
add|adl
commit|cSmmit
push|Qush
status|stitus
remote|vemote
checkout|OheckOutlq
branch|brancM
pull|pkllKejto
merge|mQrgepdD
diff|difNDr
tag|BagxjkL
log|yog
fetch|fetcYCDfBc
reset|reseNvSHmh
grep|gfepBVGe

The DistanceStringMetrics:

 | Metric                           | Correct | Wrong   |
 -------------------------|---------|----------
 SmithWatermanGotohWindowedAffine | 18      | 0
SmithWatermanGotoh               | 18      | 0
MongeElkan                       | 18      | 0
JaroWinkler                      | 18      | 0 
Levenstein                       | 18      | 0
Jaro                             | 18      | 0
SmithWaterman                    | 17      | 1
QGramsDistance                   | 16      | 2
NeedlemanWunch                   | 11      | 7
ChapmanLengthDeviation           | 4       | 14
ChapmanMeanLength                | 1       | 17
CosineSimilarity                 | 1       | 17
OverlapCoefficient               | 1       | 17 
EuclideanDistance                | 1       | 17
JaccardSimilarity                | 1       | 17
DiceSimilarity                   | 1       | 17
MatchingCoefficient              | 1       | 17
BlockDistance                    | 1       | 17

**Comment by natemcmaster** [comment]
Really interesting data. Thanks for doing some research here. It looks to me like Jaro(Winkler) or Levenstein are good candidates. I'd be okay with a PR that uses either of these.

**Comment by MadbHatter** [comment]
From wikipedia:

> Informally, the Jaro distance between two words is the minimum number of single-character transpositions required to change one word into the other.

The JaroWinkler implementation favors words that start correctly and make mistakes down the line.
e.g. comtat (favored) vs aftmit.

>    the Levenshtein distance allows deletion, insertion and substitution;
    the Damerau–Levenshtein distance allows insertion, deletion, substitution, and the transposition of two     adjacent characters;

Damerau–Levenshtein in this case seems to be the best Algorithm for this (In my opinion).
I will start writing up a test-implementation for CommandLineUtils today/tommorow.

**Comment by MadbHatter** [comment]
Hey,
I created a basic implementation over at https://github.com/MadbHatter/CommandLineUtils.
See commit https://github.com/MadbHatter/CommandLineUtils/commit/3688c340459b71d76a3edaf473f62dc99f108336.

It's not the most beautiful thing, but it's something to base things on.
Things that are missing:
- A Treshold (Should we even use one, or always give the best match? What does git do?)
- A option to disable the suggestion?
- Unit tests for some things (StringDistance.GetBestMatch)

**Comment by natemcmaster** [comment]
Some thoughts:

> A Treshold (Added)

Yes, needed. Some inputs will be just completely off. How did you pick a threshold?

> Added subcommands via flattening of the tree structure (Maybe change this to only catch direct subcommands, not sure)
> Modify the behaviour so it only conisders one ParameterType? Not sure if we want this?

I think it would be best to only make suggestions would be correct if not for the typo. For example, I would consider this a bad suggestion:
```
$ git fethc
command 'fethc' not found
Did you mean `--force`?
```

> A option to disable the suggestion?

Maybe, but not essential IMO.

**Comment by MadbHatter** [comment]
> Yes, needed. Some inputs will be just completely off. How did you pick a threshold?

I normalized the distance between two strings, e.g. 1.0d-distance/length.
Meaning that 1 is a perfect match and 0 a complete nonmatch.
I then did add treshold to GetBestMatchIndex, at the moment I put it at 0.33, e.g 1 in 3 has to be right.

> I think it would be best to only make suggestions would be correct if not for the typo. For example, I would consider this a bad suggestion:

Hmm, fine I will change it.
I had more the idea making this possible:
```
myprogram copy a.exe b.exe
command copy not found
Did you mean the option --copy?
```
But I see your reasoning.

> Maybe, but not essential IMO.

Add least an option to hide the suggestion for hidden commands (Does this Library support this, I'm not sure at the moment)?

**Comment by natemcmaster** [comment]
> Did you mean the option --copy?

I like this idea. Sorry, maybe wasn't clear. What I meant was that we shouldn't make suggestions that would be incorrect. In this case, changing `copy` to `--copy` would be correct because `--copy` is an option on `myprogram`. Maybe I didn't understand what you meant by "Added subcommands via flattening of the tree structure". I thought you were implying that you might do this:
```
Usage:
myprogram subcommand [--copy]
$ myprogram --cpy
Did you mean `--copy`?
```
In this case, it would be a bad suggestion because you can't execute `myprogram --copy`, but must do `myprogram subcommand --copy`.

> Add least an option to hide the suggestion for hidden commands

Already exists. A `ShowInHelpText` property exists on commands, options, and args. If this is set to false, I would expect it to be excluded from hints.

**Comment by MadbHatter** [comment]
Sorry. I am on Mobille.

> In this case, it would be a bad suggestion because you can't execute myprogram --copy, but must do myprogram subcommand --copy.

Ah. That was not my intention but I wasn't being clear enough. At the moment it does this because I didn't have the time to fix it. My final intention is for it to be aware of valid subcommands&others only. It would also output the whole necesssary command structure as to prevent confusion. E.g
mycommand --magic vs only --magic


>Already exists. A ShowInHelpText property exists on commands,

Ah good to know. I must include this.

Thanks for your thoughts.

**Comment by jerriep** [comment]
@natemcmaster I know this is not the same thing, but are there plans to support tab completion [like the .NET CLI does](https://github.com/dotnet/cli/blob/master/Documentation/general/tab-completion.md)?

I see their commands has this thing called `WithSuggestionsFrom` which allow them to do that:
https://github.com/dotnet/cli/blob/8c937a0db08e56660aca456ac088f2d0e70735ab/src/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs#L25

Sorry for always trying to add more to your plate ;) I am building a [CLI to manage GitHub Issues](https://github.com/jerriep/github-issues-cli) and this may be a pretty nifty feature to help users quickly type the correct repo name or issue number.

**Comment by natemcmaster** [comment]
I originally opened #9 for this, but didn't get much feedback on it. No plans to implement it at the moment, but I'm open to taking a PR if someone wants to design and build it.

**Comment by jerriep** [comment]
Thanks Nate, I would be interested to look into that, but for now I want to get this GitHub CLI out of the door. I am learning a lot along the way and their are a bunch of other helper methods and utilities I think we can add.

**Comment by natemcmaster** [comment]
I've put this in the 2.3.0 milestone. @MadbHatter let me know if you are interested in contributing your prototype. It seems like a really good starting point.

**Comment by natemcmaster** [comment]
Resolved in 41553a4fb66965e90d199becdac0f771894382bf
